### PR TITLE
Define Feature Announcement component and open it on click [122]

### PR DESCRIPTION
**Pull request by samnajian**
## Description
This PR implements the Feature Announcement modal and adds in on the order detail page.

### Notes:
- Please note the eligility check is left to [another issue](https://github.com/woocommerce/shipping/issues/126)  and just to make the current PR testable some parts of it were added to this PR.
- The functionality on the frontend ( update, snooze ) will be added in [another issue](https://github.com/woocommerce/shipping/issues/123).

### Related issue(s)

https://github.com/woocommerce/shipping/issues/122

### Steps to reproduce & screenshots/GIFs
- Checkout this branch.
- Build the project.
- Set [this boolean to true](https://github.com/Automattic/woocommerce-services/pull/2743/files#diff-d4f6c48565de5a19d0ea35c131cab7b53f92578a1ac94bd9bc3f8edb6472f274R620-R622).
- Visit an order page and click on **Create shipping label** button.
- Verify the Feature Announcement modal opens up and overlaps the purchase modal.
- Verify the modal looks adheres to the Figma design mentioned on the issue.

### Checklist

- [ ] unit tests

- [ ] `changelog.txt` entry added
- [ ] `readme.txt` entry added

### CODE REVIEW

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [inline client/components/migration/feature-announcement.jsx:51]: 💅 – Cosmetic change 
The text section doesn't have full white background extended to the bottom
image

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js]: ❓️ – Explanations needed

This PR has a few of these white space/new line changes. I don't think they are necessary since WP prefers these spaces between brackets. I am not sure why `eslint` isn't complaining about it 🤷

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js]: ❔ – Curiosity – “Good to go, but I’d like to learn more.”
Is `orderId` being used? Should we remove it if it's not needed?

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [inline client/components/migration/feature-announcement.jsx]: ❔ – Curiosity – “Good to go, but I’d like to learn more.”
`useCallback` is used to optimize performance. To me, this feels like pre-mature optimization to skip re-rendering. Also, the 2nd parameter for dependencies is missing. How does React's `useCallback` if the dependency is empty? Does it always return the cache function or does it never return the cache function? How does it check when to use the cached version?

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [inline client/components/migration/feature-announcement.jsx:40]: ❔ – Curiosity – “Good to go, but I’d like to learn more.”

Should pressing `esc` not close this modal?

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [inline client/components/migration/feature-announcement.jsx:54]: 🚧 – Request changes – “It’s quite a blocking problem.”

`JSX element 'p' has no corresponding closing tag.ts(17008)` From my editor

**Review by harriswong — COMMENTED** [review]
Tested and the modal popped up. I have a few questions, LGTM otherwise.

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx:51]: This only happens on my end if the console is open i.e. the height of the page is very small, does this always happen on your end?

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js
Comment (samnajian) [reply client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js]: I was expecting eslint to check my code, but seemingly it doesn't not, I was not sure what's the preferred formatting on this repo, I'd update if I you suggest a reference we follow for this repo.

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js
Comment (samnajian) [reply client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js]: Addresses in fab5a81b3ce2c6178046cd54c90985b7e4b365ff

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx:40]: We want the users to migrate, I'd rather remove one more way of skipping the migration, but I'm happy to change this if you think we'd better allow them to use escape.
Beside using escape when 2 modals are open may get confusing.

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx:54]: the `p` tag is being closed and my IDE doesn't nag about it and it builds and works fine.

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx]: When dependencies array is left empty or not provided, the function is always cached between renders., which is fine at least for now as these functions are simply a stub and are to be implemented properly when working on https://github.com/woocommerce/shipping/issues/123

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx:51]: Yea. 
image

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx]: > When dependencies array is left empty or not provided, the function is always cached between renders.

I would like to learn more about this. Can you find me a reference please? It seems to contradict this https://react.dev/reference/react/useCallback#every-time-my-component-renders-usecallback-returns-a-different-function.

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx:40]: I am not sure. But if we want to prevent users from exiting this modal and force them to go through it. Then should we remove the "x" button at the top right?

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx:54]: Ok, then let's stick with your editor.

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx:51]: Addressed in 8629ceb98894b4e21c63f6eb2abe057d3da09f58

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx]: > If you forget the dependency array, useCallback will return a new function every time:

I was wrong!
Would it work for you if we left this to the other PR that's gonna take care of the functionality of these buttons? Hopefully the implementer of that feature doesn't make the same mistake.

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx:40]: Then they can't use the plugin's normal functionality, it will be a hard force! removing the **escape** button functionality is only removing one way of skipping the upgrade while still allowing them to close it if they want. We can ask for opinions from @MatthiasReinholz on this, or we can keep it as is for now and collect their feedback when they test it, I'm not convinced this a big problem.

**Comment by samnajian** [comment]
@harriswong Please feel free to give this PR another look if you're fine with both of the followings:
 
https://github.com/Automattic/woocommerce-services/pull/2743#discussion_r1633839387
https://github.com/Automattic/woocommerce-services/pull/2743#discussion_r1633836208

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx:40]: I agree that the `shouldCloseOnEsc` and `shouldCloseOnClickOutside` questions do not block this PR. +1 to discuss it and implement afterwards if needed.

**Review by harriswong — CHANGES_REQUESTED** [review]
🚧 – Request changes

I tested mobile view (390x844) and it looks visually squeezed together
image
image

Figma doesn't have a mobile design so I don't know what the expectation is. However, I do think it is unreadable. A quick fix could be to move the picture to the top and have the text at the bottom. Thoughts?

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx]: No problem! Thanks for checking. Then, I suggest we remove `useCallback` for the `closeModal`, `snooze` and `update` functions to something like
```js
	const closeModal = () => setIsOpen(false);

	const snooze = () => {
		// Todo: implement maybe later
	};

	const update = () => {
		// Todo: implement update
		setIsUpdating(true);
		setTimeout(() => setIsUpdating(false), 2000);
	};
```

Whoever is working on these 3 functions can then decide to add the `useCallback` should there be a need for any performance optimization. Thoughts?

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js
Comment (harriswong) [reply client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js]: `.eslintignore` ignores this folder so `eslint` won't report any errors. I don't think we need to remove the whitespace/newline because they look like they were adhering to the WordPress/Woo Standard. Removing the space seems to be a step away from the standard, which is why I am calling it out as a code reviewer.

> I was not sure what's the preferred formatting on this repo

I usually default to the WordPress/Woo standards. Ideally, `eslint --fix` will fix this for us. I don't want to come off as being nitpicky and I also don't have a strong argument given that this file is ignored in `.eslintignore`. We can check with the team or you can use your best judgement.

File: client/components/migration/feature-announcement.jsx
Comment (samnajian) [reply client/components/migration/feature-announcement.jsx]: >  should there be a need for any performance optimization

We don't always need to measure to find out about performance optimizations if:
1) We know there is a probable issue.
2) The so-to-say optimization comes with little to zero cost in terms of development time or readability, like in here.

Anyways I'm removing it for now, the other PR may add it though.

**Comment by samnajian** [comment]
> mobile

This what I see on simulated mobile anyways ( without this Feature announcement modal ) 
I can't see how the extension has been usable on mobile so far, and not sure about our policy for mobile support.

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js
Comment (samnajian) [reply client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js]: > Ideally, eslint --fix will fix this for us.

It doesn't fix the issues, not it does it for the newly added file (client/components/migration/feature-announcement.jsx) 
I remember we had a discussion about this on Feb 22nd 1Ju-42fxF1FiOymv0mrRAnJZfFsWZ3jPTGu06BGKX8Co-gdoc and we agreed to leave these things to our linters!

Comment (samnajian) [reply client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js]: The commit 2e2f50b2daf9899358328ea113ff7764566c46f5  reverts the changes in the old file, but the linter didn't do anything for the new file.

File: client/components/migration/feature-announcement.jsx
Comment (harriswong) [reply client/components/migration/feature-announcement.jsx:54]: I had a 1:1 chat with @samnajian. Adding some clarity to the issue here (discovered by my VSCode editor.

The issue here is `< /p>` gives me the following errors:
image
image

The fix is to update `< /p>` to `</p>` but I can fix that later when I work on this file.

**Review by harriswong — APPROVED** [review]
I had a 1:1 chat with @samnajian. Regarding the mobile view, we will tackle that ~will~ with a follow up.

LGTM

**Comment by harriswong** [comment]
I will merge it for you per our discussion on the call.

**Pull request merged**
